Make alps reservation query requests tunable

arungrover opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
On Cray when a job finishes PBS sends a request to ALPS a release reservation request.  PBS then will intermittently poll until the ALPS response is "No entry for resId" is received. As of today, time delay between each polling cycle is exponential in nature and the according to Cray this can be sped up.
This pull request makes the default polling interval (and jitter) small and also exposes these tunables to admins wherein they can be set appropriately.


#### Affected Platform(s)
Cray platform

#### Cause / Analysis / Design
Lowering down the default value of polling wait time to 0.4 second and polling jitter that is added to the polling wait time to 0.12 second

#### Solution Description
Adding support for new tunable "alps_release_jitter" and "alps_release_interval" which will be used to wait for each subsequent alps reservation cancellation request.

#### Testing logs/output
This functionality can only be tested on an actual Cray machine. I've added a PTL script to the pull but cray simulator is skipped because on a simulator the reservation request gets canceled when the very first cancellation request is received (unlike the real world use case).

#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [x] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [x] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [x] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.

Documentation [Link](https://pbspro.atlassian.net/wiki/spaces/PD/pages/216498179/Make+the+wait+time+between+Cray+ALPS+release+reservation+requests+adjustable)


[Craysim testlogs](https://github.com/PBSPro/pbspro/files/1956707/test_logs_from_sim.txt)


__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

lisa-altair commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        subsequest reservation cancellation requests are made atleast
+        after the set interval.
+        """
+        ari = 4.298
```
How did you arrive at this value?  Perhaps add a comment explaining.

arungrover replied:
It was just a random value big enough to measure using log message timestamps.

bhroam replied:
Put a comment explaining the value.

lisa-altair commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        subsequest reservation cancellation requests are made by adding
+        a random time interval (less than jitter) to alps_release_interval.
+        """
+        arj = 2.198
```
How did you arrive at this value? Perhaps add a comment explaining.

lisa-altair commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+            start_time = time.time()
+            j1 = Job(TEST_USER)
+            jid1 = self.server.submit(j1)
+            self.server.expect(JOB, {ATTR_state: 'R'}, id=jid1)
```
To ensure that there is enough time for the aprun to get started, maybe we should sleep for a couple seconds before we delete the job? The sleep will likely also cause the ALPS reservation not to be freed right away so that you know you are getting into the code that uses jitter.

lisa-altair commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+                    time_current = self.get_epoch(line_current)
+                    self.assertLessEqual(time_current - time_prev, max_delay,
+                                         msg="alps_release_jitter not working")
+                    time_prev = time_current
```
In this test, you don't skip if you only find one cancellation request...But shouldn't you at least keep a count and if for all 5 jobs you only find one cancellation then you should skip/error?
Which does make me wonder, is skipping the test the right thing to do?  Since we're trying to test a base functionality, wouldn't it be better to error out so someone can look into why the wait_time/jitter code isn't being executed?

arungrover replied:
Well, it gets us into a situation where we can not really say that functionality did not work. I thought for the first case it was probably ok to skip because then we can run it again or at the very least the test will not count as an error. For this test, I was not sure if I should be skipping it because there might be other retries where it might exercise the functionality.
Like you said, I can probably keep a count of retries, and skip if it did not exercise the code in any of the retries. I specifically don't want to error out because it is not PBS's fault if the first cancellation call succeeds.

lisa-altair replied:
Okay, I see what you mean.

bhroam replied:
The usual reason to skip a test is if the environment you require for the test is not available.  For example, you don't have a Cray or you need 3 moms and only have one.  The problem here is that you're testing a race condition.  If the reservation is always deleted on the first try, then PBS didn't do anything wrong.  We still didn't test what we wanted to test.  This is a hard question to answer.  If we fail, it makes it sound like PBS did something wrong.  If we skip, it makes it sound like we don't have the right environment.  Neither are the case here.  I'm leaning towards skip over fail, but I don't know what the right answer is.

arungrover commented:
@lisa-altair I've addressed your review comments and also addressed some warnings that Codacy was showing.

vccardenas commented on test/tests/functional/pbs_alps_release_tunables.py:
```diff
@@ -0,0 +1,162 @@
+        # to either run it again or find out a way to delay the reservation
+        # cancellation at ALPS level itself.
+        if len(out) >= 2:
+            _, line_prev = out[0]
```
You might want to just index the list of tuples, e.g.:
  line_prev = out[0][1]

vccardenas commented on test/tests/functional/pbs_alps_release_tunables.py:
```diff
@@ -0,0 +1,162 @@
+            _, line_prev = out[0]
+            time_prev = get_epoch(line_prev)
+            for data in out[1:]:
+                _, line_current = data
```
Similar to above comment:
    line_current = data[1]

vccardenas commented on test/tests/functional/pbs_alps_release_tunables.py:
```diff
@@ -0,0 +1,162 @@
+            # cancellation at ALPS level itself.
+            if len(out) >= 2:
+                retry -= 1
+                _, line_prev = out[0]
```
You might want to just index the list of tuples, e.g.:
  line_prev = out[0][1]

vccardenas commented on test/tests/functional/pbs_alps_release_tunables.py:
```diff
@@ -0,0 +1,162 @@
+                _, line_prev = out[0]
+                time_prev = get_epoch(line_prev)
+                for data in out[1:]:
+                    _, line_current = data
```
Similar to above comment,
    line_current = data[1]

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3730,6 +3734,46 @@ set_int(const char *id, char *value, int *var)
+	float	val;
+
+	if (value == NULL || *value == '\0') {
+		sprintf(log_buffer, "No value specified, no action taken.");
```
use snprintf()

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3730,6 +3734,46 @@ set_int(const char *id, char *value, int *var)
+	val = strtod(value, &left);
+	if (left == value || val <= 0) {
+		sprintf(log_buffer, "bad value \"%s\"", value);
+		log_event(PBSEVENT_SYSTEM, 0, LOG_ERR, id, log_buffer);
```
I'd use log_err() instead of log_event() here.

arungrover replied:
I actually retained the implementation from set_int and other setter functions.

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3730,6 +3734,46 @@ set_int(const char *id, char *value, int *var)
+	}
+	*var = val;
+
+	sprintf(log_buffer, "setting %f", val);
```
use snprintf().  Also you might consider putting more in this log message.  Maybe use id

arungrover replied:
It adds the ID as part of log_event call.

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3933,23 +3977,51 @@ set_vnode_per_numa_node(char *value)
+	ret = set_float(__func__, value, &tmp);
+	if (ret == HANDLER_SUCCESS) {
+		fract = modf(tmp, &integ);
+		alps_release_interval = (int)integ*1000000 + (fract*1000000);
```
Isn't this a usecond_t?

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3933,23 +3977,51 @@ set_vnode_per_numa_node(char *value)
- * @retval 0 failure
- * @retval 1 success
+ * @retval 0 HANDLER_FAIL
+ * @retval 1 HANDLER_SUCCESS
```
You return HANDLER_FAIL or HABNDLER_SUCCESS, not 0 or 1.

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3933,23 +3977,51 @@ set_vnode_per_numa_node(char *value)
+	ret = set_float(__func__, value, &tmp);
+	if (ret == HANDLER_SUCCESS) {
+		fract = modf(tmp, &integ);
+		alps_release_jitter = (int)integ*1000000 + (fract*1000000);
```
Same as above, shouldn't this be a usecond_t?

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        TestFunctional.setUp(self)
+
+    def get_epoch(self, msg):
+        a = time.strptime(msg[:19], "%m/%d/%Y %H:%M:%S")
```
I'm not sure :19 will work.  Is the month, day, hour, and minute always 2 characters?  If not, you'll have a variable length time.  Since this is coming from a log message, you can always split on the ';'

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        self.mom.add_config({'$alps_release_interval': ari})
+
+        # submit a job and then delete it after it starts running
+        start_time = time.time()
```
You are using this in log_match().  time.time() returns a float.  You should int() it first.

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        # cancellation at ALPS level itself.
+        if len(out) >= 2:
+            nline_prev, line_prev = out[0]
+            time_prev = self.get_epoch(line_prev)
```
Looks like nline_prev is not used.  Also, if you only use line_prev once, just use out[0][1].

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+            time_prev = self.get_epoch(line_prev)
+            for data in out[1:]:
+                nline, line_current = data
+                time_current = self.get_epoch(line_current)
```
Same as above, use data[1]

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+                self.assertGreaterEqual(time_current - time_prev,
+                                        math.ceil(ari),
+                                        msg="alps_release_interval not working"
+                                        )
```
Consider merging the ) onto the line above.

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+                time_prev = time_current
+
+        else:
+            self.skipTest("Reservation cancelled without retry, Try again!")
```
Instead of skipping this test, maybe put the meat of it into a function and try running it a few times before skipping or failing the test.

arungrover replied:
I do something similar to what you are saying while testing Jitter. That I do because Jitter is random in nature and I can never be certain of how long it will sleep. In this test there is minimal jitter (120 miliseconds) so it is more certain in nature. I believe ALPS cancels in first request itself then it will get cancelled in every other try as well. I'd like to retain it as is but If you still think it makes sense to retry then I'll do that.

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        """
+        arj = 2.198
+        ari = 1
+        max_delay = (ari + math.ceil(2.198))
```
Either put arj in math.ceil() or just use 3.  I dislike using functions on literals when the return value is known up front.

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+        # we can probably try deleting a reservation a few times.
+        for each in range(5):
+            # submit a job and then delete it after it starts running
+            start_time = time.time()
```
you probably want to int() this.

bhroam commented on test/tests/functional/pbs_alps_request_tunables.py:
```diff
@@ -0,0 +1,150 @@
+            # cancellation at ALPS level itself.
+            if len(out) >= 2:
+                nline_prev, line_prev = out[0]
+                time_prev = self.get_epoch(line_prev)
```
as above, use out[0][1]

arungrover commented:
@bhroam  and @vccardenas  Thanks for the review. I've addressed your comments.
@bhroam  I think you reviewed an old commit, some of the comments were resolved in the new commit. Please have a look again.
Thanks!

bhroam commented on src/resmom/catch_child.c:
```diff
@@ -2404,9 +2404,7 @@ del_job_hw(job *pjob)
-			for (i=1;
-				(total_time = end_time - begin_time) < alps_release_timeout;
-				++i) {
+			for (i = 1; (total_time = end_time - begin_time) < alps_release_timeout; ++i) {
```
Don't set variables in the conditional portion of a loop.  In this case, I don't see total_time being used, so you can just remove it.

arungrover replied:
Well, this part of the code is not something I have changed but I'll be happy to amend it if it creates a problem. I do see total time being used to log the amount of time we waited for canceling a reservation, it is outside the loop. Since the code has been there for years and has not caused any warnings or troubles I'd argue to have it as is, but, if you feel strongly about not having to set a variable in the conditional portion, I'll move it.

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3980,8 +3979,8 @@ set_vnode_per_numa_node(char *value)
  *
- * @retval 0 HANDLER_FAIL
- * @retval 1 HANDLER_SUCCESS
+ * @retval HANDLER_FAIL
```
Usually the retval statements say what you are returning and when it will be returned.  In your case, it's the return value of set_float().  Maybe replace these two with "@return return value of set_float()" or something like that.

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -4006,8 +4005,8 @@ set_alps_release_wait_time(char *value)
- * @retval 0 HANDLER_FAIL
- * @retval 1 HANDLER_SUCCESS
+ * @retval HANDLER_FAIL
+ * @retval HANDLER_SUCCESS
```
Same as above, use a more descriptive return

bhroam commented on test/tests/functional/pbs_alps_release_tunables.py:
```diff
@@ -92,16 +93,14 @@ def test_alps_release_wait_time(self):
+            time_prev = get_epoch(out[0][1])
             for data in out[1:]:
-                _, line_current = data
+                line_current = data[1]
```
You only use line_current once, just use data[1] there instead.

bhroam replied:
I don't see how it is more readable.  The only thing it provides is the name line_current.  If you want more explanation, use a comment.

arungrover replied:
Well, you are right, a comment would have helped but I feel it is not complex enough that it would need a comment and it is also not clear enough that data[1] is the line where log_match() finds a hit. I thought the variable name does help here. 
I'd like to know if @lisa-altair and @vccardenas also think it should be removed then I'll remove it.
I personally like to be lenient with variable use in PTL script because I feel it does not save us much to have it optimized and everyone has their own personal style of dealing with things, but I may be wrong :)

arungrover commented:
@bhroam I've addressed your comments and also made an additional change to the for loop in catch_child.c where end_time is set inside the for statement. Please have a look again.

@lisa-altair, @vccardenas Thanks for your sign-off. There has been another update to the pull, please have a look again.

vccardenas commented:
PTL tests still look good to me.

bhroam commented on src/resmom/catch_child.c:
```diff
@@ -2438,7 +2437,7 @@ del_job_hw(job *pjob)
 					"Cancelled ALPS reservation %ld after a "
 					"total of %d tries",
-					pjob->ji_extended.ji_ext.ji_reservation, i);
+					pjob->ji_extended.ji_ext.ji_reservation, i+1);
```
Add spaces around the '+'

arungrover replied:
will do

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -3979,8 +3979,7 @@ set_vnode_per_numa_node(char *value)
  *
- * @retval HANDLER_FAIL
- * @retval HANDLER_SUCCESS
+ * @return return value of set_float()
```
'return' is duplicated.  I'd remove it.

bhroam commented on src/resmom/mom_main.c:
```diff
@@ -4005,8 +4004,7 @@ set_alps_release_wait_time(char *value)
  *
- * @retval HANDLER_FAIL
- * @retval HANDLER_SUCCESS
+ * @return return value of set_float()
```
Same as above, remove the duplicate 'return'

vccardenas commented:
The PTL changes look good to me.

arungrover commented:
@bhroam I've squashed and rebased my branch. Thanks!

bhroam commented:
I merged in the PR.
